torchrl/data/utils.py

```diff
@@ -35,6 +42,98 @@
 
 
+def consolidate_spec(
+    spec: TensorSpec,
+    recurse_through_entries: bool = True,
+    recurse_through_stack: bool = True,
+):
+    """Given a TensorSpec, removes exclusive keys by adding 0 shaped specs."""
+    spec = spec.clone()
+
+    if not isinstance(spec, (CompositeSpec, LazyStackedCompositeSpec)):
+        return spec
+
+    if isinstance(spec, LazyStackedCompositeSpec):
+        keys = set(spec.keys())  # shared keys
+        lazy_keys_per_spec = [
+            set() for _ in range(len(spec._specs))
+        ]  # list of exclusive keys per td
+        lazy_keys_examples = {}  # set of all exclusive keys with an example for each
+        for spec_index in range(len(spec._specs)):  # gather all lazy keys
+            sub_spec = spec._specs[spec_index]
+            if recurse_through_stack:
+                sub_spec = consolidate_spec(
+                    sub_spec, recurse_through_entries, recurse_through_stack
+                )
+                spec._specs[spec_index] = sub_spec
+            for sub_spec_key in sub_spec.keys():
+                if sub_spec_key not in keys:  # lazy key
+                    lazy_keys_per_spec[spec_index].add(sub_spec_key)
+                    if sub_spec_key not in lazy_keys_examples:
+                        value = sub_spec[sub_spec_key]
+                        lazy_keys_examples.update({sub_spec_key: value})
+
+        for spec_index in range(len(spec._specs)):  # add missing exclusive entries
+            sub_spec = spec._specs[spec_index]
+            for lazy_key in set(lazy_keys_examples.keys()).difference(
+                lazy_keys_per_spec[spec_index]
+            ):
+                lazy_key_example = lazy_keys_examples[lazy_key]
+                sub_spec.set(
+                    lazy_key,
+                    _empty_like_spec(lazy_key_example, sub_spec.shape),
+                )
+            spec._specs[spec_index] = sub_spec
+
+    if recurse_through_entries:
+        for key in spec.keys():
+            value = spec[key]
+            if isinstance(value, (CompositeSpec, LazyStackedCompositeSpec)):
+                spec.set(
+                    key,
+                    consolidate_spec(
+                        value, recurse_through_entries, recurse_through_stack
+                    ),
+                )
+    return spec
+
+
+def _empty_like_spec(spec, shape):
+    if isinstance(spec, (CompositeSpec, LazyStackedCompositeSpec)):
+        return spec.empty()
+    elif isinstance(spec, LazyStackedTensorSpec):
+        shape = list(shape[: spec.stack_dim]) + list(shape[spec.stack_dim + 1 :])
+        return torch.stack(
+            [_empty_like_spec(sub_spec, shape) for sub_spec in spec._specs],
+            spec.stack_dim,
+        )
+    else:
+        spec_shape = spec.shape
+        shape = [dim if i < len(shape) else 0 for i, dim in enumerate(spec_shape)]
+        spec = spec[(0,) * len(spec_shape)]
```

Comment:
All dims are 0?
Why not just the first one, or the heterogeneous one?
Might be too much work or too opinionated but for a het tensor implementation having less het dimensions is always better

Reply:
> if it was the first one: why the first one?

could be convention motivated by efficiency purposes

> if it is the het: what are the het here? this key does not exist, we are creating it. all the features are het

more like:
If all dims of the tensors that do exist are the same, take the first. If at least one tensor has a different shape along one dim, take that one instead. That heuristic minimises the number of heterogeneous dims and maximises efficiency. I think taking the first dim has little side effects (that I can see) since at the end of the day the `numel()` is still 0. Having all 0s will be cause major computational burden on the het tensors if we go down the road we're on...

Reply:
when thinking of the semantics though, in the example above, you would give the user
```
 "obs": torch.tensor(2,4,-1,6, 7)
```
 imagine that we have a shape filter (the one we discussed) where we ask for shape (..., 7) for a neural network, then this filter will not filter out (2,4,0,6,7). 

 i think it is too opinionated to choose to keep 6 and 7 and not 5. and it will be ultra confusing for users.

Isn't it possible to work around the computational burden in het tensor?

Reply:
that makes things much more complex. I do really think that the lower the number of heterogeneous shapes the better. 
Wasn't the idea of the filter was to select the shapes such that they could be stacked: if a tensor in a list has 0 elements it can't be stacked. If the idea is to cat it can be cat. I don't think that this is a blocker.
Using the first dim is also convenient as the trailing dimensions are more restrictive for most usages (eg linear or conv layers).
To me making more dims than needed 0 has the risk of hiding things: if I have 3 tensordicts the first 2 with key "foo" and value shapes [1, 2], [1, 3] and no "foo" for the third, users will see a shape [-1, -1] for "foo" when (to me) [1, -1] would be more reasonable. If you call `filter(..., shape=[3])` you'll get the second value.
In the case of `[3, 1]`, `[2, 1]` and nothing, `filter(..., shape=[3, 1])` will also work seamlessly.

Reply:
ok what about all 0s when there are no het dims (instead of arbitrarily choosing the first) and only 0s in the het dims if there are?

cc @acciorocketships can I also get what you think?

Reply:
Oh yeah that seems reasonable enough

Reply:
on it